File: C++/Day13.cpp
#include "AH.h"

namespace Day13
{

	typedef std::pair<int, int> Point;
	typedef std::pair<bool, int> Fold;

	std::set<Point> paperFold(const std::set<Point> ps, const Fold f)
	{
		std::set<Point> newPs;
		for (auto p : ps) 
		{
			Point pNew;
			if (f.first)
				pNew = std::make_pair(p.first, (p.second > f.second) ? 2*f.second - p.second : p.second);
			else
				pNew = std::make_pair((p.first > f.second) ? 2*f.second - p.first : p.first, p.second);

			newPs.insert(pNew);
		}

		return newPs;
	}

	void PrintGrid(const std::set<Point> ps)
	{
		for (int i = 0; i < 6; ++i)
		{
			for (int j = 0; j < 40; ++j)
			{
				auto p = std::make_pair(j,i);
				if (ps.count(p) != 0)
					std::cout << "#";
				else
					std::cout << " ";
			}
			std::cout << std::endl;
		}
	}

	int Run(const std::string& filename)
	{
		auto inputLines = AH::ReadTextFile(filename);

		std::set<Point> points;
		std::vector<Fold> folds;

		size_t l = 0;
		while (inputLines[l].size() > 0)
		{
			const auto ps = AH::Split(inputLines[l], ',');
			const auto p = std::make_pair(std::stoi(ps[0]),std::stoi(ps[1]));
			points.insert(p);
			++l;
		}
		++l;
		for (; l < inputLines.size(); ++l)
		{
			auto str = inputLines[l];
			const bool b = (str.at(11) == 'y');
			str.erase(0,13);
			const auto f = std::make_pair(b, std::stoi(str));
			folds.push_back(f);
		}
		auto part1 = paperFold(points, folds.front());
		std::set<Point> part2 = points;
		for (auto f : folds)
			part2 = paperFold(part2, f);

		AH::PrintSoln(13, part1.size(), part2.size());
		PrintGrid(part2);
		return 0;
	}

}

File: Go/Day13.go
package main

import AH "./adventhelper"

import (
	"fmt"
	"strconv"
	"strings"
)

type Point struct {
	x, y int
}

type Fold struct {
	axis  bool
	value int
}

func uniqueCount(intSlice []Point) (count int) {
	keys := make(map[Point]bool)
		for _, entry := range intSlice {
			if _, value := keys[entry]; !value {
				keys[entry] = true
				count += 1
			}
		}
	return count
}

func parseInput(ss []string) (map[Point]bool, []Fold) {
	points := make(map[Point]bool)
	folds := []Fold{}

	group := 0;
	for i, l := range ss {
		if (len(l) > 10) {
			break
		}
		s := strings.Split(l, ",")
		l, _ := strconv.Atoi(s[0])
		r, _ := strconv.Atoi(s[1])

		points[Point{x:l, y:r}] = true
		group = i
	}

	for i := group + 1; i < len(ss); i++ {
		str := ss[i]
		b := str[11] == 'y'
		tail := str[13:]
		v, _ := strconv.Atoi(tail)
		folds = append(folds, Fold{axis:b, value:v})
	}	
	return points, folds
}

func doFold(ps map[Point]bool, f Fold) map[Point]bool {
	newPoints := make(map[Point]bool)
	for p, _ := range ps {
		var pNew Point
		if (f.axis) {
			if (p.y > f.value) {
				pNew = Point{p.x, 2*f.value - p.y}
			} else {
			pNew = Point{p.x, p.y}
			}
		} else {
			if (p.x > f.value) {
				pNew = Point{2*f.value - p.x, p.y}
			} else {
			pNew = Point{p.x, p.y}
			}			
		}
		newPoints[pNew] = true
	}

	return newPoints
}

func printGrid(ps map[Point]bool) {
	for i := 0; i < 6; i++ {
		for j := 0; j < 40; j++ {
			p := Point{j,i}
			if _, value := ps[p]; value {
				fmt.Print("#")
			} else {
				fmt.Print(" ")
			}
		}
		fmt.Println("")
	}
}

func main() {
	js, _ := AH.ReadStrFile("../input/input13.txt")

	points, folds := parseInput(js)

	part2 := points
	for _, f := range folds {
		part2 = doFold(part2, f)
	}

	AH.PrintSoln(13, len(doFold(points, folds[0])), 0)
	printGrid(part2)

	return
}
